src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -0,0 +1,687 @@
+                return root_splitted(oc, get_laddr(), get_node_meta().depth, p_tuple);
+              } else {
+                return oc.tm.dec_ref(oc.t, get_laddr())
+                  .safe_then([p_tuple = std::move(p_tuple)] (auto ret) {
+                  return merge_entry_ret(
+                         merge_entry_ertr::ready_future_marker{},
+                         mutation_result_t(mutation_status_t::SPLITTED, p_tuple, std::nullopt));
+                });
+              }
+            });
+          });
+        }
+      });
+    }
+  });
+
+}
+
+OMapInnerNode::internal_iterator_t
+OMapInnerNode::get_containing_child(const std::string &key)
+{
+  for (auto i = iter_begin(); i != iter_end(); ++i) {
+    if (i.contains(key))
+      return i;
+  }
+  ceph_assert( 0 == "invalid");
+  return iter_end();
+}
+
+std::ostream &OMapLeafNode::print_detail_l(std::ostream &out) const
+{
+  return out << ", size=" << get_size()
+         << ", depth=" << get_meta().depth;
+}
+
+OMapLeafNode::get_value_ret
+OMapLeafNode::get_value(omap_context_t oc, const std::string &key)
+{
+  logger().debug("{}: {} key = {}","OMapLeafNode", __func__, key);
+  auto ite = find_string_key(key);
+  if (ite != iter_end()) {
+    auto value = ite->get_string_val();
+    return get_value_ret(
+      get_value_ertr::ready_future_marker{},
+      std::make_pair(key, value));
+  } else {
+    return get_value_ret(
+      get_value_ertr::ready_future_marker{},
+      std::make_pair(key, ""));
+  }
+}
+
+OMapLeafNode::insert_ret
+OMapLeafNode::insert(omap_context_t oc, const std::string &key, const std::string &value)
+{
+  logger().debug("{}: {}, {} -> {}","OMapLeafNode", __func__, key, value);
+  if (!extent_will_overflow(key.size() + 1, value.size() + 1)) {
+    if (!is_pending()) {
+      auto mut = oc.tm.get_mutable_extent(oc.t, this)->cast<OMapLeafNode>();
+      return mut->insert(oc, key, value);
+    }
+    auto replace_pt = find_string_key(key);
+    if (replace_pt != iter_end()) {
+      journal_leaf_update(replace_pt, key, value, maybe_get_delta_buffer());
+    } else {
+      auto insert_pt = string_lower_bound(key);
+      journal_leaf_insert(insert_pt, key, value, maybe_get_delta_buffer());
+
+      logger().debug(
+        "{}: {} inserted {}->{} {}"," OMapLeafNode",  __func__,
+        insert_pt.get_node_key(),
+        insert_pt.get_node_val(),
+        insert_pt.get_string_val());
+    }
+    return insert_ret(
+           insert_ertr::ready_future_marker{},
+           mutation_result_t(mutation_status_t::SUCCESS, std::nullopt, std::nullopt));
+  } else {
+    return make_split_children(oc).safe_then([this, oc, &key, &value] (auto tuple) {
+      OMapNodeRef _left = std::get<0>(tuple);
+      OMapNodeRef _right = std::get<1>(tuple);
+      std::string pivot = std::get<2>(tuple);
+      OMapLeafNodeRef left = _left->cast<OMapLeafNode>();
+      OMapLeafNodeRef right = _right->cast<OMapLeafNode>();
```

Comment:
make_split_children here doesn't need to be virtual and can just return the correct types.